app/plugin_install.go

```diff
@@ -169,6 +172,59 @@ func (a *App) installPlugin(pluginFile, signature io.ReadSeeker, installationStr
 	return manifest, nil
 }
 
+// InstallMarketplacePlugin installs a plugin listed in the marketplace server. It will get the plugin bundle
+// from the prepackaged folder, if available, or remotely if EnableRemoteMarketplace is true.
+func (a *App) InstallMarketplacePlugin(request *model.InstallMarketplacePluginRequest) (*model.Manifest, *model.AppError) {
+	var pluginFile, signatureFile io.ReadSeeker
+
+	prepackagedPlugin, appErr := a.getPrepackagedPlugin(request.Id, request.Version)
+	if appErr != nil && appErr.Id != "app.plugin.marketplace_plugins.not_found.app_error" {
+		return nil, appErr
+	}
+	if prepackagedPlugin != nil {
+		fileReader, err := os.Open(prepackagedPlugin.Path)
+		if err != nil {
+			err = errors.Wrapf(err, "failed to open prepackaged plugin %s", prepackagedPlugin.Path)
+			return nil, model.NewAppError("InstallMarketplacePlugin", "app.plugin.install_marketplace_plugin.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+		pluginFile = fileReader
+		signatureFile = bytes.NewReader(prepackagedPlugin.Signature)
+	}
+
+	if *a.Config().PluginSettings.EnableRemoteMarketplace && pluginFile == nil {
```

Comment:
Isn't the remote marketplace supposed to override prepackaged if available? Maybe I don't understand how that is happening.

Reply:
It will override it if it has a newer version (and if the remote markerplace is allowed). The marketplace lists all latest versions available and when the installation is requested is done by version. If the prepackaged plugin has the latest version there is no need to fetch it from the remote source.

It will look first for the plugin in the prepackaged folder [here](https://github.com/mattermost/mattermost-server/blob/8673fcd9c918c5811baf4cd367cfc7ccd4a0b887/app/plugin_install.go#L180)(line 180 above) specifying the version requested. If it can't find the same version it then here it will go to the marketplace to get that newer version.

Reply:
Ah. So the version has already been determined at this point. I see.